forgerock-openbanking-uk-aspsp-as/forgerock-openbanking-uk-aspsp-as-gateway/forgerock-openbanking-uk-aspsp-as-gateway-server/src/test/java/com/forgerock/openbanking/aspsp/as/service/TppRegistrationServiceTest.java

```diff
@@ -0,0 +1,218 @@
+import com.forgerock.openbanking.aspsp.as.configuration.ForgeRockDirectoryConfiguration;
+import com.forgerock.openbanking.aspsp.as.configuration.OpenBankingDirectoryConfiguration;
+import com.forgerock.openbanking.common.services.store.tpp.TppStoreService;
+import com.forgerock.openbanking.constants.OpenBankingConstants;
+import com.forgerock.openbanking.exceptions.OBErrorException;
+import com.forgerock.openbanking.jwt.services.CryptoApiClient;
+import com.forgerock.openbanking.model.oidc.OIDCRegistrationRequest;
+import com.nimbusds.jose.shaded.json.JSONObject;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.AssertionsForClassTypes.catchThrowableOfType;
+import static org.mockito.Mockito.*;
+
+@RunWith(MockitoJUnitRunner.class)
+public class TppRegistrationServiceTest {
+
+    CryptoApiClient cryptoApiClient = mock(CryptoApiClient.class);
+    OpenBankingDirectoryConfiguration openBankingDreictoryConfiguration = mock(OpenBankingDirectoryConfiguration.class);
+    ForgeRockDirectoryConfiguration forgeRockDirectoryConfiguration = mock(ForgeRockDirectoryConfiguration.class);
+    TppStoreService tppStoreService = mock(TppStoreService.class);
+    AMOIDCRegistrationService amoidcRegistrationService = mock(AMOIDCRegistrationService.class);
+    TppEntriesKPIService tppEntriesKPIService = mock(TppEntriesKPIService.class);
+    TppRegistrationService tppRegistrationService = null;
+
+    OIDCRegistrationRequest registrationRequest = mock(OIDCRegistrationRequest.class);
+    List<String> validRedirectUris = new ArrayList <>(List.of("https://google.co.uk"));
+    String softwareId = "SoftwareID";
+    JWTClaimsSet validSsaJwtClaims = null;
+
+
+    @Before
+    public void setUp() throws Exception {
+        this.tppRegistrationService = new TppRegistrationService(this.cryptoApiClient,
+            this.openBankingDreictoryConfiguration, this.forgeRockDirectoryConfiguration,
+            this.tppStoreService, this.amoidcRegistrationService, this.tppEntriesKPIService);
+
+
+        this.validSsaJwtClaims = new JWTClaimsSet.Builder()
+            .claim(OpenBankingConstants.SSAClaims.SOFTWARE_ID, softwareId)
+            .claim(OpenBankingConstants.SSAClaims.SOFTWARE_REDIRECT_URIS, this.validRedirectUris)
+            .build();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        reset(cryptoApiClient, openBankingDreictoryConfiguration, forgeRockDirectoryConfiguration, tppStoreService,
+                amoidcRegistrationService, tppStoreService, registrationRequest);
+    }
```

Comment:
Using the above approach means you don't need to reset any injected mocks.

Reply:
Added as a task in https://github.com/OpenBankingToolkit/openbanking-aspsp/issues/320